### PR TITLE
Fix back to start

### PR DESCRIPTION
**Pull request by SebastianSvanland**
Added the BackToStart module as a ROS node in ATOS. Includes functionality for resetting all test objects to their starting position and reload scenario trajectory. 
Need to move path publishers from BackToStart module to objectControl since the trajectory will only update in the gui when calculating the back-to-start trajectory and not when reloading the original trajectory.

### CODE REVIEW

File: modules/BackToStart/CMakeLists.txt
Comment (Robert108) [inline modules/BackToStart/CMakeLists.txt]:
```suggestion
```

File: modules/BackToStart/CMakeLists.txt
Comment (Robert108) [inline modules/BackToStart/CMakeLists.txt]:
```suggestion
```

File: modules/BackToStart/CMakeLists.txt
Comment (Robert108) [inline modules/BackToStart/CMakeLists.txt]:
```suggestion
```

File: modules/BackToStart/inc/backtostart.hpp
Comment (Robert108) [inline modules/BackToStart/inc/backtostart.hpp]:
```suggestion
```

File: modules/BackToStart/src/backtostart.cpp
Comment (Robert108) [inline modules/BackToStart/src/backtostart.cpp]: Since this is a a new major module when should create tests for it so that we fulfill our own guidelines :) 
You can take a look at https://github.com/RI-SE/ATOS/pull/599 for inspiration.

File: modules/BackToStart/src/backtostart.cpp
Comment (Robert108) [inline modules/BackToStart/src/backtostart.cpp]:
```suggestion
```

File: modules/BackToStart/src/backtostart.cpp
Comment (Robert108) [inline modules/BackToStart/src/backtostart.cpp]:
```suggestion
```

File: modules/BackToStart/src/backtostart.cpp
Comment (Robert108) [inline modules/BackToStart/src/backtostart.cpp]:
```suggestion

```

Comment (Robert108) [reply modules/BackToStart/src/backtostart.cpp]: Double check (hint. Maybe with a test) or remove

File: common/module.hpp
Comment (samuelthoren) [inline common/module.hpp]: I don't think `Reload Settings` is a clear enough name, because we only reset the trajectory, or am I wrong? Maybe it could be named something like `Reset Trajectory` making it more clear what it does.

File: gui/controlpanel/public/javascripts/client.js
Comment (samuelthoren) [inline gui/controlpanel/public/javascripts/client.js:77]: Might be out of scope for this PR, but pressing this button crashes object control. We should maybe not make it possible to press this button until this issue is fixed.

File: gui/controlpanel/views/index.pug
Comment (samuelthoren) [inline gui/controlpanel/views/index.pug:26]: This might be hard to do, but is it possible to only show these buttons only when launching the experimental nodes? Because these buttons don't do anything when launching ATOS "normally", which could lead to confusion.

File: launch/launch_experimental.py
Comment (samuelthoren) [inline launch/launch_experimental.py]:
```suggestion
```
This row isn't necessary unless you want to read parameters from `params.yaml`.

File: modules/BackToStart/CMakeLists.txt
Comment (samuelthoren) [inline modules/BackToStart/CMakeLists.txt]: Should the commented rows be here, otherwise remove them.

File: modules/BackToStart/inc/backtostart.hpp
Comment (samuelthoren) [inline modules/BackToStart/inc/backtostart.hpp]:
```suggestion
```

File: modules/BackToStart/src/backtostart.cpp
Comment (samuelthoren) [inline modules/BackToStart/src/backtostart.cpp]:
```suggestion
void BackToStart::onExitMessage(const ROSChannels::Exit::message_type::SharedPtr) {
```

File: modules/BackToStart/src/backtostart.cpp
Comment (samuelthoren) [inline modules/BackToStart/src/backtostart.cpp]: Remove the code you've commented out, also the one at the bottom of the file if you shouldn't keep it.

File: modules/BackToStart/src/main.cpp
Comment (samuelthoren) [inline modules/BackToStart/src/main.cpp]: Does this need to be static?

Comment (samuelthoren) [reply modules/BackToStart/src/main.cpp]: Use camelCase for variable names!

File: modules/ObjectControl/inc/objectcontrol.hpp
Comment (samuelthoren) [inline modules/ObjectControl/inc/objectcontrol.hpp]: Don't forget TODO!

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (samuelthoren) [inline modules/ObjectControl/src/objectcontrol.cpp:212]: I noticed that it is the same command here and at `onReloadSettingsMessage`, is this correct?

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (samuelthoren) [inline modules/ObjectControl/src/objectcontrol.cpp:813]: This doesn't seem to work correctly, when I press `Abort` -> `Clear Abort` -> `Reset Test`, the back to start trajectory shows up for only a short time in the 3D-view, but then the original trajectory is shown again. For the map view however, it's working correctly.

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (samuelthoren) [inline modules/ObjectControl/src/objectcontrol.cpp:765]: Add documentation to all functions

**Review by samuelthoren — CHANGES_REQUESTED** [review]
Good work! It's nice to have this functionality back. I have added some comments, the biggest issue is that the back to start trajectory doesn't seem to show up correctly in the 3D-view. It shows up for about a second, and then I get the original trajectory again.

File: modules/EsminiAdapter/src/esminiadapter.cpp
Comment (Robert108) [inline modules/EsminiAdapter/src/esminiadapter.cpp:687]: Nice!

File: common/module.hpp
Comment (Robert108) [reply common/module.hpp:77]: Agreed

File: gui/controlpanel/views/index.pug
Comment (Robert108) [reply gui/controlpanel/views/index.pug:26]: Yes, We should add in any case a condition to the buttons that they are not clickable in curtain states. Maybe the BTS buttons can listen for msgs being published that are specific to the BTS module.

**Review by Robert108 — CHANGES_REQUESTED** [review]
Great to have BTS available again! Remember to add Doxygen comments and ReadTheDocs documentation.

File: common/module.hpp
Comment (SebastianSvanland) [reply common/module.hpp]: The idea here is that the reload settings is supposed to reload everything that is needed to run the test again, it just so happens that it is only the trajectory that has been changed during the reset but could potentially be more so I wanted to keep it open. However, I do agree that these names (reset test included) are difficult for a user to understand and that maybe it should be done automatically when the object has returned to its starting position so you only need to click on one button for everything (reset test in this case which would reload settings as well). Though that would require a lot more functionality with triggers on positions etc which isn't really implemented in ATOS yet. 

Let me know if you still feel like it needs a name change

File: gui/controlpanel/public/javascripts/client.js
Comment (SebastianSvanland) [reply gui/controlpanel/public/javascripts/client.js:77]: Hmm, I don't experience this. Could it be in combination with something else?

File: gui/controlpanel/views/index.pug
Comment (SebastianSvanland) [reply gui/controlpanel/views/index.pug:26]: This is the case for many buttons if pressed in the incorrect order. I would suggest a new task to upgrade the GUI with this kind of functionality

File: modules/BackToStart/CMakeLists.txt
Comment (SebastianSvanland) [reply modules/BackToStart/CMakeLists.txt]: Oops thought I had removed them when I noticed everything worked

File: modules/BackToStart/src/main.cpp
Comment (SebastianSvanland) [reply modules/BackToStart/src/main.cpp]: I think the static was remainder from copying some code, doesn't need to be there so I removed it. Also changed to camelCase

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (SebastianSvanland) [reply modules/ObjectControl/src/objectcontrol.cpp:212]: They are both related to the reset functionality but I changed it to `COMM_OSEM`, should hopefully be more accurate?

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (SebastianSvanland) [reply modules/ObjectControl/src/objectcontrol.cpp:813]: This functionality actually behaves as it should. However, if you run the experimental launch file with trajectorylet_streamer it will bug out. I'm guessing that the trajectorylet_streamer continues to publish trajectories on the topic.

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (SebastianSvanland) [reply modules/ObjectControl/src/objectcontrol.cpp:765]: I added some documentation in the hpp-file, do you think it's enough? I noticed there is no documentation for the rest of the code.

**Comment by SebastianSvanland** [comment]
Can't merge until #599 is merged.

**Comment by SebastianSvanland** [comment]
I'm not sure what else to write about the module in the readthedocs, the module itself is very simple. Any suggestions @Robert108 or is it enough?

File: modules/BackToStart/tests/test_backtostart.cpp
Comment (Robert108) [inline modules/BackToStart/tests/test_backtostart.cpp:113]: Could you assert that the last point in the back to start traj is actually near the start point in the original traj?

File: docs/Usage/Modules/BackToStart.md
Comment (Robert108) [inline docs/Usage/Modules/BackToStart.md]: "The service is only allowed in `Disarmed` state and calling `Reset test` in the GUI does not result in a state change."
 Sorry i don't understand this. A step by step guide on how the function is used after a test would be awesome!

**Review by Robert108 — CHANGES_REQUESTED** [review]

File: docs/Usage/Modules/BackToStart.md
Comment (SebastianSvanland) [reply docs/Usage/Modules/BackToStart.md]: I included a step by step guide as well as two images showing what happens when pressing `Reset Test`, hopefully it's more clear now.

File: common/module.hpp
Comment (samuelthoren) [reply common/module.hpp]: Yes, that's a good point. We can maybe keep the name as it is for now.

File: gui/controlpanel/public/javascripts/client.js
Comment (samuelthoren) [reply gui/controlpanel/public/javascripts/client.js:77]: It happen if I press `Init` -> `Connect` -> `Disconnect`, and I'm using one object. Maybe we can take a look at this together, because it seems to happen everytime for me.

File: gui/controlpanel/views/index.pug
Comment (samuelthoren) [reply gui/controlpanel/views/index.pug:26]: I agree, we should make a ticket for this. Maybe the buttons that can't be pressed should be greyed out or something like that.

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (samuelthoren) [reply modules/ObjectControl/src/objectcontrol.cpp:813]: I see, I think that `TrajectoryletStreamer` will continue publishing the original trajectory. I can add this as a bug, because `TrajectoryletStreamer` should also publish the back to start trajectory.

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (samuelthoren) [reply modules/ObjectControl/src/objectcontrol.cpp:765]: A lot of the old code doesn't have much documentation, but for the newer code we have documented the code in the `.cpp`-files. See for instance `OSIAdapter` or `EsminiAdapter`.

Comment (SebastianSvanland) [reply modules/ObjectControl/src/objectcontrol.cpp:765]: Added some more documentation, the functions are pretty straight forward in my opinion.

Comment (samuelthoren) [reply modules/ObjectControl/src/objectcontrol.cpp:765]: I meant Doxygen comments for the function descriptions. Comments inside the functions are not really necessary for the most parts.

File: docs/Usage/Modules/BackToStart.md
Comment (victorjarlow) [inline docs/Usage/Modules/BackToStart.md]:
```suggestion
This module provides a service that enables resetting test objects to their starting position after a completed test.
The service calculates Williamson turns on both ends of the original trajectory, which if followed by the test object will lead it back to the start of the original trajectory, with a correct heading.
```

File: common/module.hpp
Comment (victorjarlow) [inline common/module.hpp]: Could maybe be called ResetTestObjectsMessage

File: common/module.hpp
Comment (victorjarlow) [inline common/module.hpp]: ReloadObjectSettingsMessage

File: modules/ObjectControl/inc/objectcontrol.hpp
Comment (victorjarlow) [inline modules/ObjectControl/inc/objectcontrol.hpp]: Why is a new Resetting state needed when there is a isResetting member variable?

File: common/module.hpp
Comment (victorjarlow) [reply common/module.hpp]: Remember to change in all places if switching to this name

File: modules/ObjectControl/inc/objectcontrol.hpp
Comment (victorjarlow) [inline modules/ObjectControl/inc/objectcontrol.hpp]: Requests a new trajectory? The below method is called setObjectTrajectory. Either of comment/method name does not describe what actually happens in the method.

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (victorjarlow) [inline modules/ObjectControl/src/objectcontrol.cpp]: Could be called e.g. republishTrajectories. The GUI is just one of many potential "listeners" on the topic.

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (victorjarlow) [inline modules/ObjectControl/src/objectcontrol.cpp:819]: We need to implement a simple "republisher" node that takes CartesianTrajectories as input and republishes them as GnssPaths... But It is not part of this PR.

**Review by victorjarlow — APPROVED** [review]
Good job, I left some comments mainly relating to clarifying names. The next step in this work is to incorporate a resetting strategy (i.e. operator determines if all should reset at the same time, or object by object, and so on..)

**Review by Robert108 — COMMENTED** [review]
Looks good, I made some changes to the sample module unit tests and pushed matching changes to this PR.

File: modules/ObjectControl/inc/objectcontrol.hpp
Comment (SebastianSvanland) [reply modules/ObjectControl/inc/objectcontrol.hpp]: Remains from another approach that I forgot to remove. Fixed now

File: modules/ObjectControl/src/objectcontrol.cpp
Comment (SebastianSvanland) [reply modules/ObjectControl/src/objectcontrol.cpp]: Renamed in new commit

File: modules/ObjectControl/inc/objectcontrol.hpp
Comment (SebastianSvanland) [reply modules/ObjectControl/inc/objectcontrol.hpp]: Updated the documentation, let me know if you're satisfied

**Review by Robert108 — APPROVED** [review]

**Review by samuelthoren — APPROVED** [review]
Good job on the PR! It doesn't crash anymore when I press disconnect, however in Foxglove it still says it is connected when the object has disconnected when I press the button. Please fix this before you merge.

**Comment by SebastianSvanland** [comment]
> Good job on the PR! It doesn't crash anymore when I press disconnect, however in Foxglove it still says it is connected when the object has disconnected when I press the button. Please fix this before you merge.

Once again, I do not experience the same issues as you do. Maybe we can look at it together

**Pull request merged**
